server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java

```diff
@@ -0,0 +1,182 @@
+            );
+        }
+
+        String reasonToDecline = checkIndexMetadata(shardRouting, targetNode, allocation);
+        if (reasonToDecline != null) {
+            return allocation.decision(Decision.NO, NAME, getDecisionDetails(false, shardRouting, targetNode, reasonToDecline));
+        }
+
+        if (shardRouting.primary()) {
+            return primaryShardDecision(shardRouting, targetNode, allocation);
+        }
+        return replicaShardDecision(shardRouting, targetNode, allocation);
+    }
+
+    // handle scenarios for allocation of a new shard's primary copy
+    private Decision primaryShardDecision(ShardRouting primaryShardRouting, DiscoveryNode targetNode, RoutingAllocation allocation) {
+        if (!targetNode.isRemoteStoreNode()) {
+            return allocation.decision(Decision.NO, NAME, getDecisionDetails(false, primaryShardRouting, targetNode, ""));
+        }
+        return allocation.decision(Decision.YES, NAME, getDecisionDetails(true, primaryShardRouting, targetNode, ""));
+    }
+
+    private Decision replicaShardDecision(ShardRouting replicaShardRouting, DiscoveryNode targetNode, RoutingAllocation allocation) {
+        if (targetNode.isRemoteStoreNode()) {
+            ShardRouting primaryShardRouting = allocation.routingNodes().activePrimary(replicaShardRouting.shardId());
+            DiscoveryNode primaryShardNode = allocation.nodes().getNodes().get(primaryShardRouting.currentNodeId());
+            if (!primaryShardNode.isRemoteStoreNode()) {
+                return allocation.decision(
+                    Decision.NO,
+                    NAME,
+                    getDecisionDetails(false, replicaShardRouting, targetNode, " since primary shard copy is not yet migrated to remote")
+                );
+            }
+            return allocation.decision(
+                Decision.YES,
+                NAME,
+                getDecisionDetails(true, replicaShardRouting, targetNode, " since primary shard copy has been migrated to remote")
+            );
+        }
+        return allocation.decision(Decision.YES, NAME, getDecisionDetails(true, replicaShardRouting, targetNode, ""));
+    }
+
+    // check for remote_store_enabled indices
+    private String checkIndexMetadata(ShardRouting shardRouting, DiscoveryNode targetNode, RoutingAllocation allocation) {
+        IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+        if (IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.exists(indexMetadata.getSettings())) {
+            remoteStoreEnabledIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexMetadata.getSettings());
+        }
+        // allocations and relocations must be towards a remote node
+        if (remoteStoreEnabledIndex && !targetNode.isRemoteStoreNode()) {
+            return " because a remote_store_enabled index's shard copy can only move towards a remote_store node";
```

Comment:
`s/move/be allocated` .  no need for `_` in `remote_store_enabled` .  Also lets call it `remote store backed index.

Reply:
Updated the description.